src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DocumentPullDiagnosticsHandler.cs

```diff
@@ -0,0 +1,195 @@
+using System.Runtime.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.HtmlCSharp
+{
+    [Shared]
+    [ExportLspMethod(MSLSPMethods.DocumentPullDiagnosticName)]
+    internal class DocumentPullDiagnosticsHandler :
+        LSPProgressListenerHandlerBase<DocumentDiagnosticsParams, DiagnosticReport[]?>,
+        IRequestHandler<DocumentDiagnosticsParams, DiagnosticReport[]?>
+    {
+        private readonly LSPRequestInvoker _requestInvoker;
+        private readonly LSPDocumentManager _documentManager;
+        private readonly LSPDocumentMappingProvider _documentMappingProvider;
+        private readonly LSPProgressListener _lspProgressListener;
+
+        [ImportingConstructor]
+        public DocumentPullDiagnosticsHandler(
+            LSPRequestInvoker requestInvoker,
+            LSPDocumentManager documentManager,
+            LSPDocumentMappingProvider documentMappingProvider,
+            LSPProgressListener lspProgressListener)
+        {
+            if (requestInvoker is null)
+            {
+                throw new ArgumentNullException(nameof(requestInvoker));
+            }
+
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
+            if (documentMappingProvider is null)
+            {
+                throw new ArgumentNullException(nameof(documentMappingProvider));
+            }
+
+            if (lspProgressListener is null)
+            {
+                throw new ArgumentNullException(nameof(lspProgressListener));
+            }
+
+            _requestInvoker = requestInvoker;
+            _documentManager = documentManager;
+            _documentMappingProvider = documentMappingProvider;
+            _lspProgressListener = lspProgressListener;
+        }
+
+        // Internal for testing
+        internal async override Task<DiagnosticReport[]?> HandleRequestAsync(DocumentDiagnosticsParams request, ClientCapabilities clientCapabilities, string token, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (clientCapabilities is null)
+            {
+                throw new ArgumentNullException(nameof(clientCapabilities));
+            }
+
+            if (!_documentManager.TryGetDocument(request.TextDocument.Uri, out var documentSnapshot))
+            {
+                return null;
+            }
+
+            if (!documentSnapshot.TryGetVirtualDocument<CSharpVirtualDocumentSnapshot>(out var csharpDoc))
+            {
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Only really need to do this inbetween async operations

Reply:
I thought we wanted one of these at the start of a function (needs to be moved up) and not necessarily inbetween async operations?

Reply:
Technically callers should be doing this before hand 😄

Reply:
Ah gotcha. Removed.